### PR TITLE
Anfield feature

### PR DESCRIPTION
**Pull request by naveena-maplelabs**
Anfield: Nossql-Cassandra source support …

- Export cassandra sources and cassandra jobs data
  from source cluster.
- Import Cassandra source to target cluster, requires ssh
  credentials and database credentials of the server.
- Import Cassandra Jobs.

### CODE REVIEW

File: samples/project-anfield/import_cluster_config.py [hunk 1]
```diff
@@ -293,6 +326,50 @@ def create_sources(source, environment, node):
...
+            db_password = configparser[name].get('db_password', None)
+            if not (ssh_password and ssh_username):
+                ERROR_LIST.append(
+                    "Please provide ssh credentials for %s" % name)
+                return
+            if not (db_password and db_username):
+                ERROR_LIST.append(
+                    "Please provide database credentials for %s" % name)
+                return
+            _cassandra_params = node["registrationInfo"]["cassandraParams"]
```
Comment (pyashish) [inline samples/project-anfield/import_cluster_config.py:344]: why _ ?

File: samples/project-anfield/import_cluster_config.py [hunk 2]
```diff
@@ -293,6 +326,50 @@ def create_sources(source, environment, node):
...
+                    username=ssh_username, password=ssh_password)
+            cassandra_params["cassandraCredentials"] = dict(
+                username=db_username, password=db_password)
+            body = dict(
+                environment="kCassandra",
+                cassandraParams=cassandra_params)
+            api = 'data-protect/sources/registrations'
+            status_code, resp = rest_obj.post(
+                api, version='v2', data=json.dumps(body))
+            if status_code == 201:
```
Comment (pyashish) [inline samples/project-anfield/import_cluster_config.py:368]: if status code is non 201 , then are we putting this part of ERROR_LIST ?

File: samples/project-anfield/import_cluster_config.py [hunk 3]
```diff
@@ -745,6 +828,12 @@ def create_protection_jobs():
                                 "following protocol %s. Supported protocol is kNfs." % (
                                     job_name, name, ", ".join(protocol)))
                             break
+                if environment == "kCassandra":
```
Comment (pyashish) [inline samples/project-anfield/import_cluster_config.py:831]: make a constant KCASSANDRA at the top and use that constant in all places.

File: samples/project-anfield/rest_client.py [hunk 4]
```diff
@@ -12,40 +12,52 @@ def __init__(self, ip, username, password, domain):
...
         """
-        api = self.api_prefix + api
+        api = self.generate_api(api=api, version=version)
         headers["Authorization"] = "Bearer {}".format(self.token)
         response = requests.get(url=api, headers=headers, verify=False)
         return response.status_code, response.content
 
 
-    def put(self, api, headers={}, **kwargs):
+    def put(self, api, version='v1', headers={}, **kwargs):
```
Comment (pyashish) [inline samples/project-anfield/rest_client.py:44]: If its v1, we use sdk right ? why is it needed ?

**Review by pyashish — CHANGES_REQUESTED** [review]

File: samples/project-anfield/import_cluster_config.py [hunk 5]
```diff
@@ -178,40 +178,73 @@ def create_vaults():
...
+                nodes = sources[0].nodes
+                for node in nodes:
+                    reg_info = node["registrationInfo"]
+                    if reg_info["accessInfo"]["endpoint"] == name:
+                        status = reg_info["authenticationStatus"]
+                        break
+            # Check for the registration status, if the process is
+            # pending, sleep for 10sec and poll again.
+            if status in ["kScheduled", "kPending"]:
+                time.sleep(sleep_time*5)
```
[the remark this answers is not in the document]
Comment (naveena-maplelabs) [reply samples/project-anfield/import_cluster_config.py:205]: Done

File: samples/project-anfield/import_cluster_config.py (hunk 2 above)
Comment (naveena-maplelabs) [reply samples/project-anfield/import_cluster_config.py:368]: Incase source registration is failed we are adding the error to ERROR_LIST

File: samples/project-anfield/import_cluster_config.py (hunk 3 above)
Comment (naveena-maplelabs) [reply samples/project-anfield/import_cluster_config.py:831]: Okay

File: samples/project-anfield/rest_client.py (hunk 4 above)
Comment (naveena-maplelabs) [reply samples/project-anfield/rest_client.py:44]: We need to call v1 Api's incase of private api's

Comment (naveena-maplelabs) [reply samples/project-anfield/rest_client.py:44]: For gflags we use private v1 api.

**Pull request merged**